File: exercise/week01/php/Submarine.php
<?php

class Submarine
{
    public int $horizontal = 0;
    public int $depth = 0;

    public function execute(string $command): void
    {
        [$action, $value] = explode(' ', $command);
        $value = (int) $value;

        match ($action) {
            'forward' => $this->horizontal += $value,
            'down' => $this->depth += $value,
            'up' => $this->depth -= $value,
            default => null,
        };
    }

    public function executeAll(array $commands): void
    {
        foreach ($commands as $command) {
            if (trim($command) !== '') {
                $this->execute($command);
            }
        }
    }

    public function finalPosition(): int
    {
        return $this->horizontal * $this->depth;
    }
}

File: exercise/week01/c#/Submarine/Submarine.cs
namespace Neural
{
    public class Submarine
    {
        public int Horizontal { get; private set; } = 0;
        public int Depth { get; private set; } = 0;

        public void Execute(string command)
        {
            var parts = command.Split(' ');
            var action = parts[0];
            var value = int.Parse(parts[1]);

            switch (action)
            {
                case "forward":
                    Horizontal += value;
                    break;
                case "down":
                    Depth += value;
                    break;
                case "up":
                    Depth -= value;
                    break;
            }
        }

        //IEnumerable<string> is a more general type than string[], allowing for more flexibility in the input
        public void ExecuteAll(IEnumerable<string> commands)
        {
            foreach (var command in commands)
            {
                if (!string.IsNullOrWhiteSpace(command))
                    Execute(command);
            }
        }

        public int FinalPosition() => Horizontal * Depth;

    }
}
